app/src/main/scala/org/alephium/explorer/persistence/queries/TransactionQueries.scala

```diff
@@ -368,28 +365,29 @@ trait TransactionQueries extends StrictLogging {
                   gasPrice)
     }
 
-  def getBalanceQuerySQL(address: Address): DBActionSR[(U256, Option[TimeStamp])] = {
+  def getBalanceQueryDEPRECATED(address: Address): DBActionSR[(U256, Option[TimeStamp])] = {
     sql"""
         SELECT outputs.amount, outputs.lock_time
         FROM outputs
-        LEFT JOIN inputs ON outputs.key = inputs.output_ref_key
-        WHERE outputs.main_chain = true AND outputs.address = $address AND inputs.block_hash IS NULL
+        LEFT JOIN inputs
+        ON outputs.key = inputs.output_ref_key
+        WHERE outputs.main_chain = true
+        AND outputs.address = $address
+        AND inputs.block_hash IS NULL
       """.as[(U256, Option[TimeStamp])]
   }
 
-  val getBalanceQuery = Compiled { address: Rep[Address] =>
-    mainOutputs
-      .filter(output => output.address === address)
-      .joinLeft(mainInputs)
-      .on(_.key === _.outputRefKey)
-      .filter(_._2.isEmpty)
-      .map { case (output, _) => (output.amount, output.lockTime) }
-  }
-
-  def getBalanceAction(address: Address): DBActionR[(U256, U256)] = {
-    getBalanceQuery(address).result.map { outputs =>
-      sumBalance(outputs)
-    }
+  def getBalanceQuery(address: Address): DBActionSR[(U256, Option[TimeStamp])] = {
+    sql"""
+      SELECT outputs.amount, outputs.lock_time
+      FROM outputs
+      LEFT JOIN inputs
```

Comment:
We will need to join based on `inputs.main_chain = true`. Otherwise, an output that is spent by an orphan block will be considered as spent in this query.

To make this query safe, I'd propose to join on only `main_chain = true` for both table, i.e. `ON (outputs.key = inputs.output_ref_key AND outputs.main_chain = true AND inputs.main_chain = true)`.

For performance, we probably need to add `main_chain` index for both inputs and outputs table.

Reply:
:+1: 
`main_chain` index is already there for both table

Reply:
The change might affect the execution. We will need to check with the query planner, Gate.io address to make sure the performance is good.

Reply:
![maim-1650619694](https://user-images.githubusercontent.com/2979182/164678453-30762512-414e-4bd8-9f76-7c2be2567bf0.png)
 Here are the previous and new query plan

Reply:
I don't have all outputs updated locally so it's not yet at full speed.